### ISSUE TITLE
ENH: Vectorize stats tests

### ISSUE THREAD

**Issue by drewoldag** [scipy/scipy#17440]
### Is your feature request related to a problem? Please describe.

The current implementation of `anderson` `kstest` and `cramervonmises` do not support, and in the case of CvM, explicitly prevent vectorized calculations. 

We have a use case where it will be rather valuable to be able to perform these calculations in a vectorized way. Our code will evaluate those metrics for ~billions of instances of scipy distributions used to evaluate the performance of astronomical photometric redshift estimates. 

### Describe the solution you'd like.

Ideally, the statistical tests would be vectorized out of the box. And indeed, most of the code is vectorized, but in each of the examples listed, there's at least one line of code the breaks the vectorizability. 

### Describe alternatives you've considered.

We've contemplated implementing these tests ourselves, but contributing it to the broader community is viewed as a very positive thing. :) 

### Additional context (e.g. screenshots, GIFs)

I would like to get a conversation started to determine what would be involved in contributing to SciPy in order to implement these in a vectorized way. 

@mdhaber - I've seen your name on lots of stats-related tickets, I would love to hear your take on this. And of course, my apologies if I'm calling out the wrong person! If I am, would you be able to point me in the right direction?

**Comment by mdhaber** [comment]
The general issue is gh-14651. @tupui and I will resolve that issue as part of this project:
https://chanzuckerberg.com/eoss/proposals/scipy-fundamental-tools-for-biomedical-research/

However, that would only vectorize these tests for convenience, not speed. So there is still something to discuss here.

I think you'd be interest in gh-16967 which unifies `anderson`, `kstest`, and `cramervonmises`. Here's my takeaway: after gh-16967, there should be a followup PR that makes it easy for the user to perform additional goodness of fit tests on new data but using the same null distribution. What do you think of that?

**Comment by drewoldag** [comment]
@mdhaber I agree gh-16967 does seem to do a lot of what we're driving for, and I think a follow up PR to vectorize for speed would be really helpful for our use case. In the code that we're developing, we wrap all the rv_continuous Scipy distributions and support a few additional piecewise distributions as well. 

Honestly, my statistics knowledge is limited, so I'm going to tag in @aimalz to get some help addressing the last couple of questions in your comment.

**Comment by drewoldag** [comment]
I spoke offline to @aimalz, and he agreed with the assessment that a followup PR for performance would be welcomed. 

I also got some clarification from Alex on the expected use cases for your second question:
> Do you need to assess goodness of fit for a specific distribution (e.g. the standard normal) or a family of distributions (e.g. any scaled/shifted normal distribution)?

"While generality would be nice, our main use case for these three is the uniform distribution between 0 and 1." He'll likely jump in here with some more thoughts. 

Similarly for this question:
> Are p-values based on asymptotic approximations sufficient, or are you looking for more accurate p-values?

"We don't actually use p-values for these metrics, largely because we've never had a case where values [couldn't be assessed] more or less by eye, and also frequentist statistics like p-values aren't commonly used in astronomy." Again, Alex will probably add more detail to this is follow up comment.

**Comment by mdhaber** [comment]
Thanks! Well, in the meantime, I think these functions (copy-pasted from gh-16967) will work for you. For n-dimensional arrays, the statistic is taken along the last axis `axis=-1` (e.g. for 2d arrays, each row is treated as a separate sample).

<details>

```python3
def _anderson_darling(dist, data):
    x = np.sort(data, axis=-1)
    n = data.shape[-1]
    i = np.arange(1, n+1)
    Si = (2*i - 1)/n * (dist.logcdf(x) + dist.logsf(x[..., ::-1]))
    S = np.sum(Si, axis=-1)
    return -n - S


def _compute_dplus(cdfvals):  # adapted from _stats_py before gh-17062
    n = cdfvals.shape[-1]
    return (np.arange(1.0, n + 1) / n - cdfvals).max(axis=-1)


def _compute_dminus(cdfvals, axis=-1):
    n = cdfvals.shape[-1]
    return (cdfvals - np.arange(0.0, n)/n).max(axis=-1)


def _kolmogorov_smirnov(dist, data):
    x = np.sort(data, axis=-1)
    cdfvals = dist.cdf(x)
    Dplus = _compute_dplus(cdfvals)  # always works along last axis
    Dminus = _compute_dminus(cdfvals)
    return np.maximum(Dplus, Dminus)


def _cramer_von_mises(dist, data):
    x = np.sort(data, axis=-1)
    n = data.shape[-1]
    cdfvals = dist.cdf(x)
    u = (2*np.arange(1, n+1) - 1)/(2*n)
    w = 1 / (12*n) + np.sum((u - cdfvals)**2, axis=-1)
    return w

```

</details>

**Comment by drewoldag** [comment]
This looks really great! I'll pull the new code an give it a shot, thanks a bunch!

**Comment by drewoldag** [comment]
@mdhaber I've pulled the lastest code and built it locally, one thing that I'm finding is that the dimensionality of input variable `data` is checked in [_fit.py:_gof_iv](https://github.com/scipy/scipy/blob/main/scipy/stats/_fit.py#L1240-L1243), which guards against input data with more than 1 dimension. I wasn't expecting that behavior, given that the code comments seem to indicate that multi dimensional data is supported. 

Am I misunderstanding the implementation? 

In our use case we would likely call stats.goodness_of_fit like so:
```
output = stats.goodness_of_fit(stats.norm, p.rvs(size=100, random_state=9999), statistic='ad')
```

Here `p` is a collection of N scipy distributions, and `p.rvs(size=M)` will return a numpy array of size NxM. 

So, of course, when p.rvs(...) is passed to goodness_of_fit, `ValueError: 'data' must be a one-dimensional array of numbers.` is returned.

**Comment by josef-pkt** [comment]
I don't see a reason why code should support vectorizing for `p is a collection of N scipy distributions`

There is no shortcut to taking a loop `np.asarray([stats.xxx(pi) for p in p_list])` AFAIR

**Comment by drewoldag** [comment]
That's a fair point @josef-pkt, and I don't have the foundation yet to know where the bottlenecks would be, but my intuition is that it would be more efficient to evaluate `goodness_of_fit` on the results of `np.asarray([stats.xxx(pi) for p in p_list])` as opposed to calling `goodness_of_fit` for each iteration of that loop. 

Especially given that the implementations of `_anderson_darling`, `_kolmogorov_smirnov`, and `_cramer_von_mises` all seem to support 2D `data`. 

But if that's way off, please let me know! :)

**Comment by josef-pkt** [comment]
for a specific distribution, cdf and other moments are vectorized. That means functions that call those can use vectorized data.

However, there is no vectorization across distributions 
`[distr.cdf(x) for distr in [stats.norm, stats.t]]` cannot be vectorized across different distributions, even though we can vectorize in `x`

If distributions were from the same parametric family, then it could be vectorized on distribution parameters, loc, scale, shape, also.

in the code for the gof statistics by mhaber above, cdfvals = distr.cdf(x) need to loop over distributions.

**Comment by drewoldag** [comment]
Oh, interesting, as it so happens in our case the collection of distributions are all of the same type. Which sounds like it could be vectorized by the existing code, yeah?

**Comment by mdhaber** [comment]
> I wasn't expecting that behavior

Hmm. I indicated above that I could consider vectorizing it, but that meant that it was not done yet. The functions I posted here do what you need in the meantime, right?
